### DISCUSSION TITLE
Profiling a distributed system

### DISCUSSION THREAD

**Discussion by fhalde** [async-profiler/async-profiler#711] [Q&A]
Hello everyone,

I have been using async-profiler for profiling a single JVM for a while now however I have no clue about the following distributed system scenario

We are trying to debug an issue in our Apache Spark application which processes data in a distributed manner. We are noticing that our jobs run slower on clusters of r5a.4xlarge (AWS) ( 16 core, 128GB RAM ) machines compared to clusters of r5a.2xlarge ( 8 core, 64GB RAM )

The cluster contains the same # of cores & RAM overall. i.e. a cluster of 2xl will have N machines whereas a cluster of 4xl will have N / 2 machines

Each spark job runs on an "executor" which does the data processing. On our 2xlarge clusters, 1 machine contains 1 executor however on 4xlarge machines 1 machine hosts 2 executors ( without any isolations mechanisms such as cgroups/taskset )

Each executor uses 8 core for its work and 54g heap

What we have done

We profiled individual executors in continuous mode for 2 minute durations and generated jfr files and finally aggregated them using

```
jfr assemble
```

```
./profiler.sh collect -e cpu -d 120 -i 20ms -o jfr ${pid}
```

the assembled jfr file was huge ( ~1.5gb ) and JVM mission control couldn't even load it ( even by giving it 20g heap )

We created flamegraphs from the combined jfr file and well that didn't make any sense either

What are some suggestions for debugging distributed systems?

**Comment by krzysztofslusarski** [comment]
To generate any representation of JFR (like flame graph, call tree, etc.) you usually load the whole file into memory (I don't know any viewer that would offload the work to HD). You can try different JFR viewer. I loaded 3GB of JFRs with my viewer:
https://github.com/krzysztofslusarski/jvm-profiling-toolkit
IIRC with heap size around 32GB. 

There is also a JFR to flame graph converted included in async-profiler, you can try that one. You also have one in Intellij Idea Utilmate. Mission Control is not the only one.

For distributed system there is also a problem of connecting the pars of JFR from multiple JVMs. I described it her:
- https://krzysztofslusarski.github.io/2022/12/12/async-manual.html#context-id
- https://krzysztofslusarski.github.io/2022/04/26/distributed.html

**Reply by fhalde** [reply]
interesting i'll read your article in a while. so jfr assemble is not sufficient to merge the several jfrs from each individual jvm's?

**Reply by krzysztofslusarski** [reply]
Depending on your needs. If you want to see aggregated results then it's sufficient. If you want to trace single request then no. Second thing that can be hard/impossible with ```assemble``` is single-node "failure". Consider a case where you join 100 JFRs and you spot high CPU usage in a single method that came from one file from one node. You need to have some kind of "filename" in the graph to spot it and to get info which node is to blame. Not sure if ```assemble``` can give it to you. 

For example here:
![image](https://user-images.githubusercontent.com/57556371/215254327-a8ce037d-f4f2-45df-abc1-480f1e32597d.png)
the "aqua" frames is a name of original JFR, that gives me a possibility to trace such issues.

**Reply by fhalde** [reply]
interesting, how did you get the jfr names in the flamegraphs?

**Reply by fhalde** [reply]
perhaps its already in your article

**Reply by krzysztofslusarski** [reply]
My tool has such an option -> https://github.com/krzysztofslusarski/jvm-profiling-toolkit

After loading JFR you have "Additional levels" on the left, one of additional level is "Add filename level".

**Reply by nitsanw** [reply]
> To generate any representation of JFR (like flame graph, call tree, etc.) you usually load the whole file into memory (I don't know any viewer that would offload the work to HD).

That is incorrect. The `convertor` included with `async-profiler` will read the file one event at a time and aggregate the stacks into a histogram/summary profile mapping each stack trace to the number of samples. The memory footprint will depend on the number of unique stacks, package/class/method/thread names etc but not the size of the JFR file. You read the whole file, but need not hold it in memory.

**Reply by krzysztofslusarski** [reply]
You are right, thanks for deeper explanation.
